### ISSUE TITLE
git_status expands all nodes whenever it is interacted with

### ISSUE THREAD

**Issue by braxtons12** [nvim-neo-tree/neo-tree.nvim#717]
I haven't been able to find a past issue/discussion about this, but this seems like something that should have been solved already, so please point me to the configuration setting I'm missing if that's the case.

As is, whenever the git_status is interacted with, it automatically expands all nodes in it, regardless of the previous state they were in. For example, if I git_add_file a file, the entire tree re-expands. This can be really jarring and annoying when you have things like generated docs in your project filling your git_status view, and them constantly being re-expanded every time you try to perform a git action.

**Comment by pysan3** [comment]
This is actually a feature hardcoded for example here. https://github.com/nvim-neo-tree/neo-tree.nvim/blob/70d3daa22bf24de9e3e284ac659506b1374e3ae2/lua/neo-tree/sources/buffers/lib/items.lua#L100-L102.

This behavior is actually the default for all sources except `filesystem`.
I discovered this while resolving #999.

I will add code so that it respects the `<source-name>.follow_current_file.leave_dirs_open` for all sources proposed here https://github.com/nvim-neo-tree/neo-tree.nvim/issues/999#issuecomment-1601887559, which will satisfy this feature request as a side product.

**Comment by cseickel** [comment]
The problem here is that all sources except filesystem are designed to explicitly open all folders automatically. This just males the most sense for 99% of situations.

> I will add code so that it respects the <source-name>.follow_current_file.leave_dirs_open for all sources proposed here https://github.com/nvim-neo-tree/neo-tree.nvim/issues/999#issuecomment-1601887559, which will satisfy this feature request as a side product.

That doesn't really fit here at all because this feature has no relation to `follow_current_file`. This would be a new feature for the `git_status` source explicitly which would probably be called `auto_expand_folders` or something similar and default to true. In this case, if it existed then @braxtons12 could set that to false and then it could behave like the filesystem source and start off collapsed but keep track of what was actually explicitly opened.

Alternatively, `git_status` could have a new collection called `explicitly_closed` folders where it keeps track of the opposite action and ensures that only those folders are forced closed after each refresh. That's probably more what is needed here and would not require a new option.

I don't think we should consider making any changes to any other sources, as it does not make sense at all to have closed folders for the `buffers` source.

**Comment by pysan3** [comment]
> This just makes the most sense for 99% of situations.

I actually took it a different way. I agree that sources other than `filesystem` should have all nodes expanded by default, but I think all the confusion comes from the fact that expanded and collapsed nodes change after another `:Neotree ...` or a refresh with no exact (intuitive) pattern.

I think the main feature people want (among all the related issues) is to simply leave the tree expansion as-is and don't mess around automatically on the second call.

> I will add code so that it respects the <source-name>.follow_current_file.leave_dirs_open for all sources proposed here https://github.com/nvim-neo-tree/neo-tree.nvim/issues/999#issuecomment-1601887559

I think I should revert this cuz this config option is not appropriate. Rather, I think a more global option such as `<source-name>.keep_dirs_expansion (default: false)` would be more suitable.

**Comment by cseickel** [comment]
> I think I should revert this cuz this config option is not appropriate. Rather, I think a more global option such as `<source-name>.keep_dirs_expansion (default: false)` would be more suitable.

I don't think this needs a new option, it can just be the new default behavior to remember explicitly closed dirs and keep them that way across refreshes for the git_status source. 

The only reason this is not the current default is that no one up until now has ever discussed  closing a directory in the git_status source so it never occurred to me that it could be a problem for someone. I personally have never done it.

**Comment by pysan3** [comment]
> it can just be the new default behavior to remember explicitly closed dirs and keep them that way across refreshes for the git_status source.

Nice idea. I will look into it and make a PR for this after v3.0 is released ;)

**Comment by showermat** [comment]
> I don't think we should consider making any changes to any other sources, as it does not make sense at all to have closed folders for the `buffers` source.

Heh, this is interesting because I found this issue trying to figure out why I couldn't do just that.  I use the buffer list to track files relevant to my current project scattered across a massive code base, and when I'm working on a certain aspect of a project it makes sense to hide certain subtrees so I can more quickly find files in others.  To me, it makes the most sense for every source to treat expanding and collapsing in the same way, remembering and honoring the set of nodes that the user has collapsed and expanded unless some other behavior like following the current file is requested.  But I'm sure you've thought this through much more carefully than I have.

This is definitely in the "nice-to-have" category rather than basic functionality; I'd just like to make it known that there *is* at least one person who would find persistence of collapsed and expanded nodes in the `buffers` source useful.
